sensor/common/virtualmachine/component_impl.go

```diff
@@ -0,0 +1,160 @@
+package virtualmachine
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/centralsensor"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/errox"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stackrox/rox/pkg/utils"
+	"github.com/stackrox/rox/sensor/common"
+	"github.com/stackrox/rox/sensor/common/message"
+)
+
+var (
+	errInputChanClosed   = errors.New("channel receiving virtual machines is closed")
+	errStartMoreThanOnce = errors.New("unable to start the component more than once")
+)
+
+type componentImpl struct {
+	centralReady    concurrency.Signal
+	lock            *sync.Mutex
+	stopper         concurrency.Stopper
+	toCentral       <-chan *message.ExpiringMessage
+	virtualMachines chan *storage.VirtualMachine
+}
+
+func (c *componentImpl) Capabilities() []centralsensor.SensorCapability {
+	return nil
+}
+
+func (c *componentImpl) Send(ctx context.Context, vm *storage.VirtualMachine) error {
+	if !c.centralReady.IsDone() {
+		log.Warnf("Cannot send virtual machine %q to Central because Central is not reachable", vm.GetId())
+		return errox.ResourceExhausted
+	}
+	select {
+	case <-ctx.Done():
+		return errors.Wrap(ctx.Err(), "context is done")
+	case c.virtualMachines <- vm:
+		return nil
+	}
+}
+
+func (c *componentImpl) Name() string {
+	return "virtualMachine.componentImpl"
+}
+
+func (c *componentImpl) Notify(e common.SensorComponentEvent) {
+	log.Info(common.LogSensorComponentEvent(e))
+	switch e {
+	case common.SensorComponentEventCentralReachable:
+		c.centralReady.Signal()
+	case common.SensorComponentEventOfflineMode:
+		// As clients are expected to retry virtual machine upserts when Sensor is in
+		// offline mode, there is no need to do anything here other than reset the signal.
+		c.centralReady.Reset()
+	}
+}
+
+// ProcessMessage is a no-op because Sensor does not receive any virtual machine data
+// from Central.
+func (c *componentImpl) ProcessMessage(ctx context.Context, msg *central.MsgToSensor) error {
+	return nil
+}
+
+// ResponsesC returns a channel with messages to Central. It must be called
+// after Start() for the channel to be not nil.
+func (c *componentImpl) ResponsesC() <-chan *message.ExpiringMessage {
+	c.lock.Lock()
+	defer c.lock.Unlock()
```

Comment:
[nit] would RLock work here?

Reply:
I think so, done in 1fe12d368e20f8badd3452da38c49520de962fdf